oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

```diff
@@ -215,8 +190,90 @@ private void reportProgress(String id) {
+        boolean downloadCompleted = false;
+        Map<String, Integer> exceptions = new HashMap<>();
+        this.nextLastModified = 0;
+        this.lastIdDownloaded = null;
+        while (!downloadCompleted) {
+            try {
+                if (lastIdDownloaded != null) {
+                    LOG.info("Recovering from broken connection, finishing downloading documents with _modified={}", nextLastModified);
+                    downloadRange(new DownloadRange(nextLastModified, nextLastModified + 1, lastIdDownloaded), childrenFilter);
+                    // We have managed to reconnect, reset the failure timestamp
+                    failuresStartTimestamp = null;
+                    numberOfFailures = 0;
+                    // Continue downloading everything starting from the next _lastmodified value
+                    downloadRange(new DownloadRange(nextLastModified + 1, Long.MAX_VALUE, null), childrenFilter);
+                } else {
+                    downloadRange(new DownloadRange(nextLastModified, Long.MAX_VALUE, null), childrenFilter);
+                }
+                downloadCompleted = true;
+            } catch (MongoException e) {
+                if (e instanceof MongoInterruptedException || e instanceof MongoIncompatibleDriverException) {
+                    // Non-recoverable exceptions
+                    throw e;
+                }
+                if (failuresStartTimestamp == null) {
+                    failuresStartTimestamp = Instant.now().truncatedTo(ChronoUnit.SECONDS);
+                }
+                LOG.warn("Connection error downloading from MongoDB.", e);
+                long secondsSinceStartOfFailures = Duration.between(failuresStartTimestamp, Instant.now()).toSeconds();
+                if (secondsSinceStartOfFailures > retryDuringSeconds) {
+                    // Give up. Get a string of all exceptions that were thrown
+                    StringBuilder summary = new StringBuilder();
+                    for (Map.Entry<String, Integer> entry : exceptions.entrySet()) {
+                        summary.append("\n\t").append(entry.getValue()).append("x: ").append(entry.getKey());
+                    }
+                    throw new RetryException(retryDuringSeconds, summary.toString(), e);
+                } else {
+                    numberOfFailures++;
+                    LOG.warn("Retrying download in {} ms; number of times failed: {}; current series of failures started at: {} ({} seconds ago)",
+                            retryIntervalMs, numberOfFailures, failuresStartTimestamp, secondsSinceStartOfFailures);
+                    exceptions.compute(e.getClass().getSimpleName() + " - " + e.getMessage(),
+                            (key, val) -> val == null ? 1 : val + 1
+                    );
+                    try {
+                        Thread.sleep(retryIntervalMs);
+                    } catch (InterruptedException ignore) {
+                    }
```

Comment:
Could it be a problem that we are not interrupting the thread while it's sleeping when this exception is caught? 
It seems that if an interruption actually occurs, the code would only catch it and continue what it's doing. That might be the intended behaviour, but I'm not 100 % sure. And if not could we do something like: 

```java
try {
    Thread.sleep(retryIntervalMs);
} catch (InterruptedException ignore) {
    // re-interrupt the current thread to set its interrupted status
    Thread.currentThread().interrupt();
}
```

Reply:
Yes, you are right. The method with this code already declares throwing `InterruptedException` and the caller is handling that exception explicitly, so the catch clause is not needed here. I removed it.